### PR TITLE
TELCODOCS-146 - document hyperthreading config in OCP 

### PR DESCRIPTION
**Pull request by aireilly**
https://issues.redhat.com/browse/TELCODOCS-146

Document hyperthreading, implications for latency, and how to enable or disable in OCP cluster.

### CODE REVIEW

**Comment by aireilly** [comment]
Build output: https://telcodocs-146--ocpdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#about_hyperthreading_for_low_latency_and_real_time_applications_cnf-master

https://telcodocs-146--ocpdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#configuring_hyperthreading_for_a_cluster_cnf-master

**Comment by aireilly** [comment]
@MarSik @gsr-shanks please @ anyone else that needs to review.

**Comment by aireilly** [comment]
@ahardin-rh Can you please label/tag as appropriate. This is for 4.7.

**Comment by fromanirh** [comment]
looks good. I'd mention that users who configure a performance profile first, and then enable/disable HT, must change their performance profile to match the cpu core IDs.
Most notably, disabling a previously-enabled HT may make some core ID unknown to the system, leading to an incorrect configuration, which in turn may cause the node to be unavailable.

**Comment by gsr-shanks** [comment]
@mrniranjan @dshchedr Please review if not already

**Comment by aireilly** [comment]
> looks good. I'd mention that users who configure a performance profile first, and then enable/disable HT, must change their performance profile to match the cpu core IDs.
> Most notably, disabling a previously-enabled HT may make some core ID unknown to the system, leading to an incorrect configuration, which in turn may cause the node to be unavailable.

What about these as notes: 

[NOTE]
====
If you configure a performance profile, and subsequently change the hyperthreading configuration for the host, ensure that you update the performance profile CPU `isolated` and `reserved` fields to match the new configuration.
====

[WARNING]
====
Disabling a previously-enabled host hyperthreading configuration can cause the CPU core IDs listed in the `PerformanceProfile` yaml to be incorrect. This incorrect configuration can cause the node to become unavailable because the listed CPUs can no longer be found.
====

**Comment by aireilly** [comment]
@fromanirh pushed these changes, please review: https://telcodocs-146--ocpdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#configuring_hyperthreading_for_a_cluster_cnf-master

**Comment by fromanirh** [comment]
warning and notice both LGTM.

**Comment by aireilly** [comment]
hi @fromanirh @mrniranjan commented that `noht` is not correct. the correct kernel arg is `nosmt`. I've made the change above.

**Comment by aireilly** [comment]
Hi @fromanirh, can you give me final sign off on the noht > nosmt change please? Should be good to merge then.

File: modules/cnf-about_hyperthreading_for_low_latency_and_real_time_applications.adoc [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Module included in the following assemblies:
+//
+// scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc
+
+[id="about_hyperthreading_for_low_latency_and_real_time_applications_{context}"]
+= About hyperthreading for low latency and real-time applications
+
+Hyperthreading is an Intel processor technology that allows a physical CPU processor core to function as two logical cores, executing two independent threads simultaneously. Hyperthreading allows for better system throughput for certain workload types where parallel processing is beneficial. The default {product-title} configuration expects hyperthreading to be enabled by default.
+
+For telco applications, it is important to design your application infrastructure to minimize latency as much as possible. Hyperthreading can slow performance times and negatively affect throughput for compute intensive workloads that require low latency. Disabling hyperthreading ensures predictable performance and can decrease processing times for these workloads.
```
Comment (ahardin-rh) [inline modules/cnf-about_hyperthreading_for_low_latency_and_real_time_applications.adoc:10]: s/telco/ telecommunications

File: modules/cnf-about_hyperthreading_for_low_latency_and_real_time_applications.adoc [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+[NOTE]
+====
+Hyperthreading implementation and configuration differs depending on the hardware you are running {product-title} on. Consult the relevant host hardware tuning information for more details of the hyperthreading implementation specific to that hardware. Disabling hyperthreading can increase the cost per core of the cluster.
+====
+
+.Additional resources
+
+* xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#configuring_hyperthreading_for_a_cluster_{context}[Configuring hyperthreading for a cluster]
```
Comment (ahardin-rh) [inline modules/cnf-about_hyperthreading_for_low_latency_and_real_time_applications.adoc:19]: At this time, we cannot include xrefs in modules, only assemblies :sob:

File: modules/configuring_hyperthreading_for_a_cluster.adoc [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+To configure hyperthreading for an {product-title} cluster, set the CPU threads in the performance profile to the same cores that are configured for the reserved or isolated CPU pools.
+
+[NOTE]
+====
+If you configure a performance profile, and subsequently change the hyperthreading configuration for the host, ensure that you update the CPU `isolated` and `reserved` fields in the `PerformanceProfile` yaml to match the new configuration.
+====
+
+[WARNING]
+====
+Disabling a previously-enabled host hyperthreading configuration can cause the CPU core IDs listed in the `PerformanceProfile` yaml to be incorrect. This incorrect configuration can cause the node to become unavailable because the listed CPUs can no longer be found.
```
Comment (ahardin-rh) [inline modules/configuring_hyperthreading_for_a_cluster.adoc:17]: s/previously-enabled/previously enabled

File: modules/configuring_hyperthreading_for_a_cluster.adoc [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+1   0    0      1    1:1:1:0       yes    4800.0000 400.0000
+2   0    0      2    2:2:2:0       yes    4800.0000 400.0000
+3   0    0      3    3:3:3:0       yes    4800.0000 400.0000
+4   0    0      0    0:0:0:0       yes    4800.0000 400.0000
+5   0    0      1    1:1:1:0       yes    4800.0000 400.0000
+6   0    0      2    2:2:2:0       yes    4800.0000 400.0000
+7   0    0      3    3:3:3:0       yes    4800.0000 400.0000
+----
++
+In this example, there are eight logical CPU cores running on four physical CPU cores. CPU0 and CPU4 are running on physical Core0, CPU1 and CPU5 are running on physical Core 1, and so on.
```
Comment (ahardin-rh) [inline modules/configuring_hyperthreading_for_a_cluster.adoc:51]: CPU1, and CPU5

File: modules/configuring_hyperthreading_for_a_cluster.adoc [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+...
+  cpu:
+    isolated: 0-4
+    reserved: 1-5
+...
+----
+
+[IMPORTANT]
+====
+Hyperthreading is enabled by default on most Intel processors. If you enable hyperthreading, all threads processed by a particular core need to be isolated or processed on the same core.
```
Comment (ahardin-rh) [inline modules/configuring_hyperthreading_for_a_cluster.adoc:80]: s/need to be/must be

File: modules/configuring_hyperthreading_for_a_cluster.adoc [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+====
+
+[id="disabling_hyperthreading_for_low_latency_applications_{context}"]
+== Disabling hyperthreading for low latency applications
+
+When configuring clusters for low latency processing, you should consider whether you want to disable hyperthreading before you deploy the cluster. To disable hyperthreading, do the following:
```
Comment (ahardin-rh) [inline modules/configuring_hyperthreading_for_a_cluster.adoc:86]: s/you should consider/consider

File: modules/configuring_hyperthreading_for_a_cluster.adoc [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+====
+
+[WARNING]
+====
+Disabling a previously-enabled host hyperthreading configuration can cause the CPU core IDs listed in the `PerformanceProfile` yaml to be incorrect. This incorrect configuration can cause the node to become unavailable because the listed CPUs can no longer be found.
+====
+
+.Prerequisites
+
+* Access to the cluster as a user with the cluster-admin role.
```
Comment (ahardin-rh) [inline modules/configuring_hyperthreading_for_a_cluster.adoc:22]: `cluster-admin` role

**Comment by ahardin-rh** [comment]
@aireilly Just some minor comments from me. You'll also have to squash you commits down to one before we can merge. Thanks!

File: modules/configuring_hyperthreading_for_a_cluster.adoc (hunk 4 above)
Comment (aireilly) [reply modules/configuring_hyperthreading_for_a_cluster.adoc:51]: This is incorrect I think Ashley. "CPU0 and CPU4 are running on physical Core0" is a clause, and "CPU1 and CPU5 are running on physical Core 1" is a clause.  Adding a comma after CPU1 breaks up the clause improperly.

**Comment by aireilly** [comment]
@ahardin-rh  I implemented your review comments, thanks. Ready for merge I guess.

File: modules/configuring_hyperthreading_for_a_cluster.adoc [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+// scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc
+
+[id="configuring_hyperthreading_for_a_cluster_{context}"]
+= Configuring hyperthreading for a cluster
+
+To configure hyperthreading for an {product-title} cluster, set the CPU threads in the performance profile to the same cores that are configured for the reserved or isolated CPU pools.
+
+[NOTE]
+====
+If you configure a performance profile, and subsequently change the hyperthreading configuration for the host, ensure that you update the CPU `isolated` and `reserved` fields in the `PerformanceProfile` yaml to match the new configuration.
```
Comment (ahardin-rh) [inline modules/configuring_hyperthreading_for_a_cluster.adoc:12]: s/yaml/YAML file

File: modules/configuring_hyperthreading_for_a_cluster.adoc [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+====
+
+[WARNING]
+====
+Disabling a previously enabled host hyperthreading configuration can cause the CPU core IDs listed in the `PerformanceProfile` yaml to be incorrect. This incorrect configuration can cause the node to become unavailable because the listed CPUs can no longer be found.
```
Comment (ahardin-rh) [inline modules/configuring_hyperthreading_for_a_cluster.adoc:17]: s/yaml/YAML file

File: modules/configuring_hyperthreading_for_a_cluster.adoc [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+----
++
+.Example output
++
+[source,terminal]
+----
+0-4
+----
+
+. Apply the isolated and reserved CPUs in the `PerformanceProfile` yaml. For example, you could set logical cores CPU0 and CPU4 as isolated, and logical cores CPU1 and CPU5 as reserved:
```
Comment (ahardin-rh) [inline modules/configuring_hyperthreading_for_a_cluster.adoc:67]: s/yaml/YAML file

**Comment by ahardin-rh** [comment]
@aireilly just one more style suggestion and we should be set. Thanks! Sorry I missed that in my 1st review.

**Pull request merged**